src/types/allocator.h

```diff
@@ -0,0 +1,98 @@
+#ifndef ALLOCATOR
+#define ALLOCATOR
+
+#include <atomic>
+#include <cstddef>
+#include <utility>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <system_error>
+
+
+class MemoryArea {
+  private:
+    std::size_t area_size = 0;
+    void* area = nullptr;
+    int descriptor;         // file descriptor
+    unsigned int offset;    // offset in the file
+
+  public:
+    MemoryArea(std::size_t size, int descrp = -1, unsigned int offset = 0)
+    :area_size(size), descriptor(descrp), offset(offset) {
+      area = ::mmap(nullptr,                      // memory area where to allocate (nullptr if it does not matter)
+                    size,                         // bytes to be mapped
+                    PROT_READ | PROT_WRITE,       // memory protection level
+                    MAP_PRIVATE | MAP_ANONYMOUS,  // mapping attributes
+                    descriptor,
+                    offset);
+      if(area == MAP_FAILED)
+      {
+        throw std::bad_alloc();
+      }
+
```

Comment:
Looks good

maybe explain the last part (mainly ::mprotect) of the constructor through a comment:
```c
    auto tail = data() + size - ::getpagesize();
    auto returnCode = ::mprotect(tail, ::getpagesize(), PROT_NONE);
    if(returnCode)
    {
        ::munmap(area, area_size);
        throw std::system_error(std::error_code(errno, std::system_category()),
                            "Protection of memory mapping [mprotect()] failed");
    }
```